### DISCUSSION TITLE
用Spark DataFrame查询doris数据并创建临时表，再使用临时表数据时，过滤条件无效

### DISCUSSION THREAD

**Discussion by sky1229329146** [apache/doris#14507] [A - General / Q&A]
用Spark DataFrame查询doris数据并创建临时表，再使用临时表数据时，过滤条件无效，而是和直接从doris拉取的数据一样
请问是什么原因呢？
有什么解决办法吗？

**Comment by gnehil** [comment]
方便举个例子吗？还有 doris 和 spark-connector 的版本也说明一下吧

**Reply by sky1229329146** [reply]
一：用的是这个依赖
<dependency>
            <groupId>org.apache.doris</groupId>
            <artifactId>spark-doris-connector-2.3_2.11</artifactId>
            <version>1.1.0</version>
        </dependency>

二：
dori版本：1.1.1
spark版本：2.4.0

三：代码如下
step1.查询限制某个条件的数据，
  val frame = DorisUtil.getDorisSession(PropertiesUtil.getPropertiesByKey("DORIS_TABLE_NAME_CAR"),"id='51100254001310801150'")

step2.打印看下符合当前条件的数据
frame.show()

step3.创建临时表
frame.createOrReplaceTempView("temptable")

step4.在临时表中限定不是这个id,居然还是能查到其他id的数据
    sparkSession.sql(
      s"""
         |select
         |*
         |from temptable
         |where id !='51100254001310801150'
         |""".stripMargin).show()
		 
		 
备注：如果在step1的下面添加 frame.cache()	,则step4中数据为空

**Reply by LemonLiTree** [reply]
frame.createOrReplaceTempView("temptable")直接去doris查询看数据确实是有其他id吗？

**Reply by sky1229329146** [reply]
是的

**Reply by gnehil** [reply]
我按照你这个步骤操作了一下，没有复现出来，方便加个联系方式帮你看一下吗

**Reply by gnehil** [reply]
或者你贴一下DorisUtil.getDorisSession这部分代码

**Reply by sky1229329146** [reply]
def getDorisSession(tableName:String,queryWhere:String="",queryFileds:String = PropertiesUtil.getPropertiesByKey("QueryFiles")) = {
    SparkUtil.getSparkSession(this.getClass.getSimpleName)
      .read
      .format("doris")
      .option("doris.table.identifier",tableName )
      .option("doris.read.field",s"$queryFileds" )
      .option("doris.filter.query",s"$queryWhere" ) 
      .option("doris.fenodes",PropertiesUtil.getPropertiesByKey("DORIS_FE_RESFUL_PORT"))
      .option("user",PropertiesUtil.getPropertiesByKey("DORIS_USERNAME"))
      .option("password",PropertiesUtil.getPropertiesByKey("DORIS_PASSWORD"))
      .option("batch_size","4096") //每次进行批量计算的行数
      .option("enable_vectorized_engine","true") //向量化执行引擎;建表时设置NOT NULL或者特殊值取代null,防止性能劣化
      .load()
  }

**Reply by gnehil** [reply]
代码看起来没有什么问题，设置一下日志等级为debug，日志搜一下 Query SQL Sending to Doris FE 看看

**Reply by sky1229329146** [reply]
我们查看debug日志发现，好像是因为我们在查询doris时使用option("doris.filter.query",s"$queryWhere" )预过滤后，再用where条件过滤，发现doris不会使用预过滤的数据而是之间将where重新作为全量数据的过滤条件，导致之前的预先过滤失效；

**Reply by gnehil** [reply]
这块确实有点问题，提个issue吧

**Reply by sky1229329146** [reply]
好的
已提交，#14781
